forte/orbital-helpers/embedding_density.cc

```diff
@@ -0,0 +1,253 @@
+    D2bb("pqrs") -= D1b("ps") * D1b("qr");
+
+    D2ab("pqrs") = D1a("pr") * D1b("qs");
+
+    psi::outfile->Printf("\n D2 formed.");
+
+    // 3-RDM
+    if (options_->get_str("THREEPDC") != "ZERO") {
+        auto D3aaa = ambit::Tensor::build(ambit::CoreTensor, "D3aaa", std::vector<size_t>(6, na));
+        auto D3aab = ambit::Tensor::build(ambit::CoreTensor, "D3aab", std::vector<size_t>(6, na));
+        auto D3abb = ambit::Tensor::build(ambit::CoreTensor, "D3abb", std::vector<size_t>(6, na));
+        auto D3bbb = ambit::Tensor::build(ambit::CoreTensor, "D3bbb", std::vector<size_t>(6, na));
+
+        D3aaa("pqrstu") += D1a("ps") * D1a("qt") * D1a("ru");
+        D3aaa("pqrstu") -= D1a("ps") * D1a("rt") * D1a("qu");
+        D3aaa("pqrstu") -= D1a("qs") * D1a("pt") * D1a("ru");
+        D3aaa("pqrstu") += D1a("qs") * D1a("rt") * D1a("pu");
+        D3aaa("pqrstu") -= D1a("rs") * D1a("qt") * D1a("pu");
+        D3aaa("pqrstu") += D1a("rs") * D1a("pt") * D1a("qu");
+
+        D3bbb("pqrstu") += D1b("ps") * D1b("qt") * D1b("ru");
+        D3bbb("pqrstu") -= D1b("ps") * D1b("rt") * D1b("qu");
+        D3bbb("pqrstu") -= D1b("qs") * D1b("pt") * D1b("ru");
+        D3bbb("pqrstu") += D1b("qs") * D1b("rt") * D1b("pu");
+        D3bbb("pqrstu") -= D1b("rs") * D1b("qt") * D1b("pu");
+        D3bbb("pqrstu") += D1b("rs") * D1b("pt") * D1b("qu");
+
+        D3aab("pqrstu") += D1a("ps") * D1a("qt") * D1b("ru");
+        D3aab("pqrstu") -= D1a("qs") * D1a("pt") * D1b("ru");
+
+        D3abb("pqrstu") += D1a("ps") * D1b("qt") * D1b("ru");
+        D3abb("pqrstu") -= D1a("ps") * D1b("rt") * D1b("qu");
+
+        psi::outfile->Printf("\n D3 formed.");
+
+        return RDMs(D1a, D1b, D2aa, D2ab, D2bb, D3aaa, D3aab, D3abb, D3bbb);
+    } else {
+        psi::outfile->Printf("\n D3 skipped.");
+
+        return RDMs(D1a, D1b, D2aa, D2ab, D2bb);
+    }
+}
+
+RDMs EMBEDDING_DENSITY::cas_rdms(std::shared_ptr<MOSpaceInfo> mo_space_info_active) {
+    // Return RDM from a CASCI/CASSCF computation
+    RDMs ref_rdms;
+    int print_level = 0;
+    if (options_->get_str("FRAGMENT_DENSITY") == "CASCI") {
+        std::string ci_type = options_->get_str("ACTIVE_SPACE_SOLVER");
+        auto state_map = to_state_nroots_map(state_weights_map_);
+
+        size_t nactv = mo_space_info_->size("ACTIVE");
+        auto actv_mos = mo_space_info_active->corr_absolute_mo("ACTIVE");
+        std::vector<int> actv_sym = mo_space_info_active->symmetry("ACTIVE");
+        auto rdocc_mos = mo_space_info_active->corr_absolute_mo("RESTRICTED_DOCC");
+
+        auto fci_ints =
+            std::make_shared<ActiveSpaceIntegrals>(ints_, actv_mos, actv_sym, rdocc_mos);
+        fci_ints->set_active_integrals_and_restricted_docc();
+
+        auto active_space_solver = make_active_space_solver(
+            ci_type, state_map, scf_info_, mo_space_info_active, fci_ints, options_);
+        active_space_solver->set_print(print_level);
+        active_space_solver->compute_energy();
```

Comment:
I wonder if this class actually do this computation or it should only handle the RDMs, which is passed in to the class or function.